### PR TITLE
[cocoa] NPE with multi-line Text control in a Group control

### LINKED ISSUE #472: [Mac] NPE using a multi-line Text control in a Group control at 100% DPI with font size
**Describe the bug**
Opening the Secure Storage Preference page in a child Eclipse on Mac M1 Ventura leads to this:

eclipse.buildId=4.25.0.I20220831-1800
java.version=17.0.5
java.vendor=Eclipse Adoptium
BootLoader constants: OS=macosx, ARCH=aarch64, WS=cocoa, NL=en_GB
Framework arguments:  -product org.eclipse.sdk.ide
Command-line arguments:  -product org.eclipse.sdk.ide -data /Users/phillipus/Documents/workspace/../runtime-New_configuration -dev file:/Users/phillipus/Documents/workspace/.metadata/.plugins/org.eclipse.pde.core/New_configuration/dev.properties -os macosx -ws cocoa -arch aarch64 -consoleLog

!ENTRY org.eclipse.jface 4 2 2022-11-24 23:15:55.749
!MESSAGE Problems occurred when invoking code from plug-in: "org.eclipse.jface".
!STACK 0
java.lang.NullPointerException: Cannot invoke "org.eclipse.swt.internal.cocoa.NSWindow.contentView()" because "window" is null
	at org.eclipse.swt.widgets.Control.calculateVisibleRegion(Control.java:775)
	at org.eclipse.swt.widgets.Control.getVisibleRegion(Control.java:2060)
	at org.eclipse.swt.widgets.Control.isObscured(Control.java:2341)
	at org.eclipse.swt.widgets.Scrollable.scrollClipViewToPoint(Scrollable.java:347)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:6754)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSend(Native Method)
	at org.eclipse.swt.internal.cocoa.NSScrollView.setDocumentView(NSScrollView.java:91)
	at org.eclipse.swt.widgets.Scrollable.setZOrder(Scrollable.java:396)
	at org.eclipse.swt.widgets.Control.createWidget(Control.java:1026)
	at org.eclipse.swt.widgets.Scrollable.createWidget(Scrollable.java:167)
	at org.eclipse.swt.widgets.Text.createWidget(Text.java:596)
	at org.eclipse.swt.widgets.Control.<init>(Control.java:126)
	at org.eclipse.swt.widgets.Scrollable.<init>(Scrollable.java:76)
	at org.eclipse.swt.widgets.Text.<init>(Text.java:140)
	at org.eclipse.equinox.internal.security.ui.storage.TabPassword.<init>(TabPassword.java:186)
	at org.eclipse.equinox.internal.security.ui.storage.StoragePreferencePage.createContents(StoragePreferencePage.java:50)
	at org.eclipse.jface.preference.PreferencePage.createControl(PreferencePage.java:244)
	at org.eclipse.jface.preference.PreferenceDialog.createPageControl(PreferenceDialog.java:1433)
	at org.eclipse.jface.preference.PreferenceDialog$8.run(PreferenceDialog.java:1196)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:174)
	at org.eclipse.jface.preference.PreferenceDialog.showPage(PreferenceDialog.java:1188)
	at org.eclipse.ui.internal.dialogs.FilteredPreferenceDialog.showPage(FilteredPreferenceDialog.java:630)
	at org.eclipse.jface.preference.PreferenceDialog$5.lambda$0(PreferenceDialog.java:660)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:74)
	at org.eclipse.jface.preference.PreferenceDialog$5.selectionChanged(PreferenceDialog.java:657)
	at org.eclipse.jface.viewers.StructuredViewer$3.run(StructuredViewer.java:821)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:174)
	at org.eclipse.jface.viewers.StructuredViewer.firePostSelectionChanged(StructuredViewer.java:818)
	at org.eclipse.jface.viewers.StructuredViewer.setSelection(StructuredViewer.java:1661)
	at org.eclipse.jface.viewers.TreeViewer.setSelection(TreeViewer.java:1090)
	at org.eclipse.jface.viewers.Viewer.setSelection(Viewer.java:376)
	at org.eclipse.ui.internal.dialogs.FilteredPreferenceDialog.setCurrentPageId(FilteredPreferenceDialog.java:446)
	at org.eclipse.ui.internal.dialogs.FilteredPreferenceDialog.openPage(FilteredPreferenceDialog.java:428)
	at org.eclipse.ui.dialogs.PreferenceLinkArea.lambda$0(PreferenceLinkArea.java:65)
	at org.eclipse.swt.events.SelectionListener$1.widgetSelected(SelectionListener.java:84)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4577)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1524)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1547)
	at org.eclipse.swt.widgets.Widget.sendSelectionEvent(Widget.java:1597)
	at org.eclipse.swt.widgets.Link.textView_clickOnLink_atIndex(Link.java:806)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:6801)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSendSuper(Native Method)
	at org.eclipse.swt.widgets.Widget.callSuper(Widget.java:236)
	at org.eclipse.swt.widgets.Widget.mouseDownSuper(Widget.java:1147)
	at org.eclipse.swt.widgets.Widget.mouseDown(Widget.java:1139)
	at org.eclipse.swt.widgets.Control.mouseDown(Control.java:2625)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:6263)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSendSuper(Native Method)
	at org.eclipse.swt.widgets.Widget.callSuper(Widget.java:236)
	at org.eclipse.swt.widgets.Widget.windowSendEvent(Widget.java:2162)
	at org.eclipse.swt.widgets.Shell.windowSendEvent(Shell.java:2511)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:6387)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSendSuper(Native Method)
	at org.eclipse.swt.widgets.Display.applicationSendEvent(Display.java:5635)
	at org.eclipse.swt.widgets.Display.applicationProc(Display.java:5774)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSend(Native Method)
	at org.eclipse.swt.internal.cocoa.NSApplication.sendEvent(NSApplication.java:117)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3936)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at org.eclipse.ui.internal.handlers.ShowPreferencePageHandler.execute(ShowPreferencePageHandler.java:58)
	at org.eclipse.ui.internal.handlers.HandlerProxy.execute(HandlerProxy.java:283)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:97)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:58)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:317)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:251)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:173)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:156)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:488)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:485)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:213)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:195)
	at org.eclipse.e4.ui.workbench.renderers.swt.cocoa.CocoaUIHandler.runCommand(CocoaUIHandler.java:421)
	at org.eclipse.e4.ui.workbench.renderers.swt.cocoa.CocoaUIHandler.access$0(CocoaUIHandler.java:407)
	at org.eclipse.e4.ui.workbench.renderers.swt.cocoa.CocoaUIHandler$1.widgetSelected(CocoaUIHandler.java:281)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4577)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1524)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1547)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1532)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1325)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4363)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3939)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:643)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:550)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:171)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)

**To Reproduce**
- Launch child Eclipse
- Open Eclipse Preferences in child Eclipse
- Try to open the General -> Security -> Secure Storage Preference page

**Environment:**
1. Select the platform(s) on which the behavior is seen:
- - [ ] All OS
- - [ ] Windows
- - [ ] Linux
- - [x] macOS Ventura Mac M1

### PR DESCRIPTION
**Pull request by SyntevoAlex**
Fixes #472
Please refer to commit message for details.

### CODE REVIEW

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Control.java
Comment (sratz) [inline bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Control.java:770]: private?

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Control.java
Comment (sratz) [inline bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Control.java:782]: calling `window()` creates a new object each time.

There's probably no race condition here (returning `null` again the 2nd time called), but maybe extract this to a local variable and only call it once, anyways?

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Control.java
Comment (SyntevoAlex) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Control.java]: Done, thanks

**Comment by SyntevoAlex** [comment]
I intend to merge this around 2023-04-15 if there're no objections.

**Comment by elsazac** [comment]
I tried this out in Mac OS (aarch64) Ventura 13.2.1

Was able to recreate the issue with the snippet and also verified the same using PR.The PR fixes the problem.So this seems good to me.
Attaching the stack trace obtained using snippet.

```
Exception in thread "main" java.lang.NullPointerException: Cannot invoke "org.eclipse.swt.internal.cocoa.NSWindow.contentView()" because "window" is null
	at org.eclipse.swt.widgets.Control.calculateVisibleRegion(Control.java:775)
	at org.eclipse.swt.widgets.Control.getVisibleRegion(Control.java:2060)
	at org.eclipse.swt.widgets.Control.isObscured(Control.java:2341)
	at org.eclipse.swt.widgets.Scrollable.scrollClipViewToPoint(Scrollable.java:347)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:6811)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSend(Native Method)
	at org.eclipse.swt.internal.cocoa.NSScrollView.setDocumentView(NSScrollView.java:91)
	at org.eclipse.swt.widgets.Scrollable.setZOrder(Scrollable.java:396)
	at org.eclipse.swt.widgets.Control.createWidget(Control.java:1026)
	at org.eclipse.swt.widgets.Scrollable.createWidget(Scrollable.java:167)
	at org.eclipse.swt.widgets.Text.createWidget(Text.java:596)
	at org.eclipse.swt.widgets.Control.<init>(Control.java:126)
	at org.eclipse.swt.widgets.Scrollable.<init>(Scrollable.java:76)
	at org.eclipse.swt.widgets.Text.<init>(Text.java:140)
	at org.eclipse.swt.bugs.Bug472$1.widgetSelected(Bug472.java:31)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4646)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1524)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1547)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1532)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1325)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4413)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3989)
	at org.eclipse.swt.bugs.Bug472.main(Bug472.java:41)
```

**Review by sratz — APPROVED** [review]

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java
Comment (sratz) [inline tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java:1682]: Font objects need to be disposed, see test failure:

```
Caused by: java.lang.Error: SWT Resource was not properly disposed
	at org.eclipse.swt.graphics.Resource.initNonDisposeTracking(Resource.java:172)
	at org.eclipse.swt.graphics.Resource.<init>(Resource.java:120)
	at org.eclipse.swt.graphics.Font.<init>(Font.java:145)
	at org.eclipse.swt.tests.junit.Test_org_eclipse_swt_widgets_Text.test_issue472(Test_org_eclipse_swt_widgets_Text.java:1682)
```

**Review by sratz — CHANGES_REQUESTED** [review]

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java
Comment (SyntevoAlex) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java]: Right, thanks for noticing. I didn't notice because I already got used that some tests always fail.

**Comment by SyntevoAlex** [comment]
The review from @sratz  is blocking the merge.
I plan to merge when the review is complete.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java
Comment (iloveeclipse) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java]: > I didn't notice because I already got used that some tests always fail.

You are not alone, see this lovely discussion at https://github.com/eclipse-platform/eclipse.platform.ui/pull/721#issuecomment-1512818194

@laeubi : this is exactly what the [broken window theory](https://en.wikipedia.org/wiki/Broken_windows_theory) explaims / predicts.

Comment (laeubi) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java]: @iloveeclipse that becomes completely hilarious, so we have an additional test failure that won't be discovered otherwise but for sure its the fault of the action. So please stop complaining and fix the "broken windows", it wasn't me that "broke" anything or started suggesting to ignoring test.

The github action even showed clearly that there are ADDITIONAL failures, so don't use that as a excuse for not following common committers responsibilities!

Comment (iloveeclipse) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java]: > So please stop complaining and fix the "broken windows

I personally believe to fix enough, so no need for an extra invite.

> don't use that as a excuse for not following common committers responsibilities!

Have you *tried* to understand https://en.wikipedia.org/wiki/Broken_windows_theory? Please read the full article, not the first few lines.

Comment (laeubi) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java]: Have you _tried_ to understand that this PR is adding a test for **macOS** that without the additional verification wont be "verified" at all? Have you _tried_ to understand that the action even **has clearly showed** that there was an additional test failure? Has you even _noticed_ that a careful reviewer even could use that information to note that there was something wrong?

> read the full article, not the first few lines.

So maybe even up to this section? 
https://en.wikipedia.org/wiki/Broken_windows_theory#Criticism


So I really don't get why you repeatably reference **me** if _other_ committers **ignore** their responsibility (either by intend or accidentally) so you probably should help them to be more productive instead of referencing the same irrelevant article again and again.

Comment (iloveeclipse) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java]: Because I hoped you would understand the problem, but I give up now.

**Review by sratz — APPROVED** [review]

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java
Comment (laeubi) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java]: @SyntevoAlex there are some long pending issues (about a year now) and I think @akurtakov has investigated some things already that there are differences between different linux distribution and/or windowing systems, so yes there is a problem but it seems the test is testing something a user won't notice at all or is irrelevant or no user complained (yet):
- GlyphMetrics: https://github.com/eclipse-platform/eclipse.platform.swt/issues/118
- Browser: https://github.com/eclipse-platform/eclipse.platform.swt/issues/103 

For `GlyphMetrics` I think there are two cases:

1. It diverge between linux systems but is okay, then we probably should simply adjust the test to accept both "good values" here.
2. It is not okay and the code should be adjusted to also work on [ubuntu-latest](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) (as of today Ubuntu 22.04) as it seems a commonly used system of Eclipse users.

For the Browser thing it seems to be an issue with that the browser modifies the HTML and the test asserts very strict, I proposed using [jsoup](https://jsoup.org/) instead (or one could try "normalize" the html manually) but no one has give any feedback on that idea yet and I don't know enough about the browser test how important the strict assert here is ... and the DND i have no idea at all and can't try it out as I usually not develop with windows.

**Comment by SyntevoAlex** [comment]
In master now.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java
Comment (SyntevoAlex) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Text.java]: I think that it would be better to make the individual Issues, and probably an umbrella Issue listing unfixed problems in tests, then split your comments here into respective Issues.

**Pull request merged**
